File: arm_nif/src/verifier.rs
use crate::error::ArmNifError;
use arm_core::instance::{AppData, Payload, Transaction};
use rustler::NifResult;

fn decode_tx(tx_bytes: &[u8]) -> Result<Transaction, ArmNifError> {
    Ok(bincode::serde::decode_from_slice(tx_bytes, bincode::config::standard())?.0)
}

/// The four payload categories (resource, encryption, external, discovery) as
/// `(blob, deletion_criterion)` lists — the boundary form crossing to Elixir as
/// `{binary, boolean}` tuples.
type AppDataBlobs = (
    Vec<(Vec<u8>, bool)>,
    Vec<(Vec<u8>, bool)>,
    Vec<(Vec<u8>, bool)>,
    Vec<(Vec<u8>, bool)>,
);

fn app_data_blobs(ad: &AppData) -> AppDataBlobs {
    let conv = |ps: &[Payload]| -> Vec<(Vec<u8>, bool)> {
        ps.iter()
            .map(|p| (p.data.clone(), p.deletion_criterion))
            .collect()
    };
    (
        conv(&ad.resource_payload),
        conv(&ad.encryption_payload),
        conv(&ad.external_payload),
        conv(&ad.discovery_payload),
    )
}

#[rustler::nif(schedule = "DirtyCpu")]
fn verify_transaction(tx_bytes: Vec<u8>) -> NifResult<bool> {
    Ok(decode_tx(&tx_bytes)?.verify().is_ok())
}

/// Decode + verify once, returning the consumed/created resources (each as
/// `(tag, app_data_blobs)`) and the consumed roots. Errors on decode failure or
/// any verification failure (carrying the specific `ArmError`).
#[rustler::nif(schedule = "DirtyCpu")]
fn verify_and_extract(
    tx_bytes: Vec<u8>,
) -> NifResult<(
    Vec<(Vec<u8>, AppDataBlobs)>,
    Vec<(Vec<u8>, AppDataBlobs)>,
    Vec<Vec<u8>>,
)> {
    let tx = decode_tx(&tx_bytes)?;
    tx.verify().map_err(ArmNifError::from)?;
    let consumed_with_appdata = tx
        .units
        .iter()
        .flat_map(|u| u.action_instance.consumed.iter())
        .map(|c| (c.nullifier.to_vec(), app_data_blobs(&c.app_data)))
        .collect();
    let created_with_appdata = tx
        .units
        .iter()
        .flat_map(|u| u.action_instance.created.iter())
        .map(|c| (c.commitment.to_vec(), app_data_blobs(&c.app_data)))
        .collect();
    let roots = tx.roots().into_iter().map(|r| r.to_vec()).collect();
    Ok((consumed_with_appdata, created_with_appdata, roots))
}

#[rustler::nif(schedule = "DirtyCpu")]
fn transaction_nullifiers(tx_bytes: Vec<u8>) -> NifResult<Vec<Vec<u8>>> {
    Ok(decode_tx(&tx_bytes)?
        .nullifiers()
        .into_iter()
        .map(|n| n.to_vec())
        .collect())
}

#[rustler::nif(schedule = "DirtyCpu")]
fn transaction_commitments(tx_bytes: Vec<u8>) -> NifResult<Vec<Vec<u8>>> {
    Ok(decode_tx(&tx_bytes)?
        .commitments()
        .into_iter()
        .map(|c| c.to_vec())
        .collect())
}

#[rustler::nif(schedule = "DirtyCpu")]
fn transaction_roots(tx_bytes: Vec<u8>) -> NifResult<Vec<Vec<u8>>> {
    Ok(decode_tx(&tx_bytes)?
        .roots()
        .into_iter()
        .map(|r| r.to_vec())
        .collect())
}

File: lib/arm_openvm/verifier.ex
defmodule ArmOpenvm.Verifier do
  # NOTE: rustler resolves `crate: :arm_nif` to `native/arm_nif/` relative to the
  # mix project root by default. In this workspace `arm_nif` lives at the repo root,
  # so set `path:` to wherever the consuming mix project can reach it, e.g.
  #   use Rustler, otp_app: :arm_openvm, crate: :arm_nif, path: "arm_nif"
  use Rustler,
    otp_app: :arm_openvm,
    crate: :arm_nif

  @moduledoc """
  NIF bindings for verifying arm-openvm resource machine transactions.

  All functions take a bincode-encoded `arm_core::instance::Transaction` as a binary.
  """

  @typedoc "Result type for NIF functions that can return errors"
  @type nif_result(t) :: t | {:error, term()}

  @typedoc "A payload blob with its deletion criterion."
  @type blob :: {binary(), boolean()}

  @typedoc "The four payload categories: {resource, encryption, external, discovery}."
  @type app_data_blobs :: {[blob()], [blob()], [blob()], [blob()]}

  @doc """
  Verify a bincode-encoded Transaction.

  Returns `true` if every action's compliance proof verifies, the revealed
  instances match, nullifiers/commitments are disjoint, and the delta proof checks;
  `false` if any of those soundness checks fail; `{:error, reason}` if the bytes
  fail to decode.
  """
  @spec verify_transaction(binary()) :: nif_result(boolean())
  def verify_transaction(_tx_bytes), do: error()

  @doc """
  Decode + verify a transaction in one pass and return its effects
  needed for global checks and storage.
  """
  @spec verify_and_extract(binary()) ::
          {[{binary(), app_data_blobs()}], [{binary(), app_data_blobs()}], [binary()]}
          | {:error, term()}
  def verify_and_extract(_tx_bytes), do: error()

  @doc "All nullifiers (32-byte binaries) in the transaction, in transaction order."
  @spec transaction_nullifiers(binary()) :: nif_result(list(binary()))
  def transaction_nullifiers(_tx_bytes), do: error()

  @doc "All commitments (32-byte binaries) in the transaction, in transaction order."
  @spec transaction_commitments(binary()) :: nif_result(list(binary()))
  def transaction_commitments(_tx_bytes), do: error()

  @doc """
  The set of consumed-resource roots in the transaction, as a deduplicated,
  sorted list of 32-byte binaries. Build a `MapSet` from it to check containment
  against the historical roots set.
  """
  @spec transaction_roots(binary()) :: nif_result(list(binary()))
  def transaction_roots(_tx_bytes), do: error()

  defp error, do: :erlang.nif_error(:nif_not_loaded)
end
